x/nft/keeper/querier.go

```diff
@@ -0,0 +1,151 @@
+			return queryOwnerByDenom(ctx, path[1:], req, k)
+		case QueryCollection:
+			return queryCollection(ctx, path[1:], req, k)
+		case QueryNFT:
+			return queryNFT(ctx, path[1:], req, k)
+		case QueryDenoms:
+			return queryDenoms(ctx, path[1:], req, k)
+		default:
+			return nil, sdk.ErrUnknownRequest("unknown nft query endpoint")
+		}
+	}
+}
+
+func queryDenoms(ctx sdk.Context, path []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+	denoms := k.GetDenoms(ctx)
+	bz := types.ModuleCdc.MustMarshalJSON(denoms)
+	return bz, nil
+}
+
+func querySupply(ctx sdk.Context, path []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+
+	var params types.QueryCollectionParams
+	err := types.ModuleCdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrUnknownRequest(sdk.AppendMsgToErr("incorrectly formatted request data", err.Error()))
+	}
+
+	collection, found := k.GetCollection(ctx, params.Denom)
+	if !found {
+		return nil, types.ErrUnknownCollection(types.DefaultCodespace, fmt.Sprintf("unknown denom %s", params.Denom))
+	}
+
+	bz := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bz, uint64(collection.Supply()))
+	return bz, nil
+}
+
+func queryOwnerByDenom(ctx sdk.Context, path []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+	var params types.QueryBalanceParams
+	err := types.ModuleCdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrUnknownRequest(sdk.AppendMsgToErr("incorrectly formatted request data", err.Error()))
+	}
+
+	var owner types.Owner
+
+	var idCollections []types.IDCollection
+	idCollection, _ := k.GetOwnerByDenom(ctx, params.Owner, params.Denom)
+	owner.Address = params.Owner
+	owner.IDCollections = append(idCollections, idCollection)
+
+	bz := types.ModuleCdc.MustMarshalJSON(owner)
+	return bz, nil
+}
+
+func queryOwner(ctx sdk.Context, path []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+	var params types.QueryBalanceParams
+	err := types.ModuleCdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrUnknownRequest(sdk.AppendMsgToErr("incorrectly formatted request data", err.Error()))
+	}
+	owner := k.GetOwner(ctx, params.Owner)
+	bz := types.ModuleCdc.MustMarshalJSON(owner)
+	return bz, nil
+}
+
+func queryCollection(ctx sdk.Context, path []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+
+	var params types.QueryCollectionParams
+	err := types.ModuleCdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrUnknownRequest(sdk.AppendMsgToErr("incorrectly formatted request data", err.Error()))
+	}
+
+	collection, found := k.GetCollection(ctx, params.Denom)
+	if !found {
+		return nil, types.ErrUnknownCollection(types.DefaultCodespace, fmt.Sprintf("unknown denom %s", params.Denom))
+	}
+
+	// collections := types.NewCollections(collection)
+	// fmt.Println("collections", collections)
+
+	bz := types.ModuleCdc.MustMarshalJSON(collection)
+	// fmt.Println("bytes collection", bz)
```

Comment:
leftover comments

Reply:
removed comments but left TODO for switching back to the marshal method you originally implemented but so it works. Or did we realize it wouldn't be possible because of the NFT type?